src/lowlevel/mpc.jl

```diff
@@ -0,0 +1,315 @@
+    ρ_max::Float64
+    separation_atol::Float64
+    separation_max::Float64
+end
+
+BooleanContact(; ρ_max=10000, separation_atol=1e-6, separation_max=100) =
+    BooleanContact(ρ_max, separation_atol, separation_max)
+
+struct LCPContact <: ContactModel
+    boolean_params::BooleanContact
+    velocity_max::Float64
+end
+
+LCPContact(; ρ_max=10000, separation_atol=1e-6, separation_max=100, velocity_max=1000) =
+    LCPContact(BooleanContact(ρ_max=ρ_max, separation_atol=separation_atol, separation_max=separation_max), velocity_max)
+
+
+function addcontact!(controller::MPCController, stage::MPCStage{C}, position::Point3D, surface::HalfSpace3D, μ::Float64, contact_model::ContactModel=BooleanContact()) where C
+    model = controller.qpmodel
+
+    normal_in_body = let state = controller.state, surface = surface, position = position
+        Parameter(model) do
+            transform(state, surface.outward_normal, position.frame)
+        end
+    end
+
+    contact_point = C(position, normal_in_body, μ, controller.state, controller.qpmodel)
+    add_contact_indicators!(controller, stage, contact_point, surface, contact_model)
+    addcontact!(stage, contact_point)
+end
+
+function add_contact_indicators!(controller::MPCController, stage::MPCStage, contact_point::ContactPoint, surface::HalfSpace3D, contact_model::BooleanContact)
+    model = controller.qpmodel
+    state = Parameter(identity, controller.state, model)
+    mechanism = controller.state.mechanism
+    position = contact_point.position
+    body = body_fixed_frame_to_body(mechanism, position.frame)
+    path_to_body = path(mechanism, root_body(mechanism), body)
+
+    current_position_world = @expression transform(state, position, root_frame(mechanism))
+    J_point = let state = state, position_world = current_position_world, path_to_body = path_to_body
+        Parameter(point_jacobian(state(), path_to_body, position_world()), model) do J
+            point_jacobian!(J, state(), path_to_body, position_world())
+        end
+    end
+    position_world = @expression current_position_world.v + stage.Δt * (J_point.J * stage.v)
+    surface_world = @expression(HalfSpace3D(
+            transform(state, surface.point, root_frame(mechanism)),
+            transform(state, surface.outward_normal, root_frame(mechanism))))
+
+    separation = @expression(dot(surface_world.outward_normal.v, position_world) -
+                             dot(surface_world.outward_normal.v, surface_world.point.v))
+
+    indicator = Variable(model)
+    @constraint(model, indicator ∈ {0, 1})
+
+    separation_min = @expression(
+        min((dot(surface_world.outward_normal.v, current_position_world.v) -
+             dot(surface_world.outward_normal.v, surface_world.point.v)),
+             -contact_model.separation_atol))
+
+    @constraint(model, [separation] >= [separation_min])
+    @constraint(model, [separation] <= [contact_model.separation_atol + contact_model.separation_max * (1 - indicator)])
```

Comment:
These can now be scalar constraints if you like.

Reply:
Scalar constraints aren't handled correctly by my half-assed LQOI patch, so I'll keep them as vectors for now